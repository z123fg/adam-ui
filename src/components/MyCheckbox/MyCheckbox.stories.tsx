import { ComponentMeta, ComponentStory} from "@storybook/react";
import React from "react";
import MyCheckbox, { CheckboxColor, CheckboxSize } from "./MyCheckbox";

export default {
  title:"Example/MyCheckbox",
  component:MyCheckbox,

} as ComponentMeta<typeof MyCheckbox>

const Template: ComponentStory<typeof MyCheckbox> = (args) => (
  <MyCheckbox {...args}/>
);

export const Small = Template.bind({})
Small.args ={
  checkboxSize:CheckboxSize.Small
}

export const Large = Template.bind({})
Large.args = {
  checkboxSize: CheckboxSize.Large
}



export const Red = Template.bind({})
Red.args = {
  checkboxColor: CheckboxColor.Red
}

export const Green = Template.bind({})
Green.args = {
  checkboxColor: CheckboxColor.Green
}

export const Blue = Template.bind({})
Blue.args = {
  checkboxColor: CheckboxColor.Blue
}

export const Disabled = Template.bind({})
Disabled.args = {
  disabled: true
}

export const DefaultCheck = Template.bind({})
DefaultCheck.args = {
  defaultCheck: true
}