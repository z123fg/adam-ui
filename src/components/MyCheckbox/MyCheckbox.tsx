import React, { FC} from "react";
import genClassName from "../../util/genClassName";

export enum CheckboxSize {
  Large = "lg",
  Small = "sm",
}


export enum CheckboxColor {
  Red = "red",
  Green = 'green',
  Blue = 'blue',
  Yellow = 'yellow'
}

interface IMyCheckboxProps {
  /**
  Different sizes for MyCheckbox
   */
  checkboxSize?: CheckboxSize;
  /**
  Different colors for MyCheckbox
   */
  checkboxColor?: CheckboxColor; 
  /**
  Diabled checkbox, cannot click it.
   */
  disabled?: boolean;
  /**
  Self-defined class name.
   */
  className?: string;
  /**
  Default checked. Can change later.
   */
  defaultCheck?:boolean;
  
}

let MyCheckbox: FC<IMyCheckboxProps>  = ({
  checkboxSize = CheckboxSize.Small,
  checkboxColor,
  disabled = false,
  className = '',
  defaultCheck = false,
  ...rest
}) => {
  let classNameStr = genClassName(
    'checkbox',
    {
      [`checkbox-${checkboxSize}`]: !!checkboxSize,
      [`checkbox-${checkboxColor}`]: !!checkboxColor,
    } , className
  ).join(' ')
  // `checkbox checkbox-${checkboxType} checkbox-${checkboxSize} checkbox-${checkboxColor}`
  let render = () => {
    if(disabled){
      return <input
        {...rest as React.DetailedHTMLProps<
          React.InputHTMLAttributes<HTMLInputElement>,
          HTMLInputElement
        >}
      disabled={disabled} 
      className={classNameStr} 
      type='checkbox' />
    }else{
      if(defaultCheck){
        return <input
          {...rest as React.DetailedHTMLProps<
            React.InputHTMLAttributes<HTMLInputElement>,
            HTMLInputElement
          >}
          defaultChecked
          className={classNameStr}
          type='checkbox' />
      }else{
        return <input 
          {...rest as React.DetailedHTMLProps<
            React.InputHTMLAttributes<HTMLInputElement>,
            HTMLInputElement
          >}
        className={classNameStr} 
        type='checkbox' />
      }
    }
  }
  
  return render()
}
export default MyCheckbox;
