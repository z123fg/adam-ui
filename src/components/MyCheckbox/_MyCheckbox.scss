.checkbox{
  position: relative;
  display: inline-block;
  color: $body-color;
  white-space: nowrap;
  text-align: center;
  vertical-align: middle;
  cursor: pointer;
}

.checkbox-sm{
  zoom:$checkbox-sm; 
}

.checkbox-lg{
  zoom:$checkbox-lg
}

.checkbox-red{
  //position: relative;
  border: 1.5px solid grey ;
  background-color: white;
  width: 13px;
  height: 13px;
  border-radius: 20%;
  -webkit-appearance:none;
  -webkit-tap-highlight-color: transparent;
  &:checked{
    content: "\2714";
    font-weight: bold;
    background-color: red;
    border: 1px solid red;
    background-clip: padding-box;
    text-align: center;
    }
  }

  .checkbox-green{

  border: 1.5px solid grey ;
  background-color: white;
  width: 13px;
  height: 13px;
  border-radius: 20%;
  -webkit-appearance:none;
  -webkit-tap-highlight-color: transparent;
  &:checked{
    content: "\2714";
    font-weight: bold;
    background-color: green;
    border: 1px solid green;
    background-clip: padding-box;
    text-align: center;
    }
  }

  .checkbox-blue{
  border: 1.5px solid grey ;
  background-color: white;
  width: 13px;
  height: 13px;
  border-radius: 20%;
  -webkit-appearance:none;
  -webkit-tap-highlight-color: transparent;
  &:checked{
    content: "\2714";
    font-weight: bold;
    background-color: rgb(40, 176, 240);
    border: 1px solid rgb(40, 176, 240);
    background-clip: padding-box;
    text-align: center;
    }
  }

  .checkbox-yellow{
  border: 1.5px solid grey ;
  background-color: white;
  width: 13px;
  height: 13px;
  border-radius: 20%;
  -webkit-appearance:none;
  -webkit-tap-highlight-color: transparent;
  &:checked{
    content: "\2714";
    font-weight: bold;
    background-color: rgb(194, 194, 28);
    border: 1px solid rgb(194, 194, 28);
    background-clip: padding-box;
    text-align: center;
    }
  }