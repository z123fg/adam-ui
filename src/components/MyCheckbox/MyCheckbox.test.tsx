import { render } from "@testing-library/react";
import MyCheckbox, { CheckboxColor, CheckboxSize } from "./MyCheckbox";

describe('test MyCheckbox Component', () => {

  it('should snapshot', () => {
    const wrapper = render(<MyCheckbox data-testid="element" />)
    const firstRender = wrapper.asFragment()
    expect(firstRender).toMatchSnapshot()
  })

  it("should render without issue", () => {
    const wrapper = render(<MyCheckbox data-testid="element" />);
    const checkboxElement = wrapper.getByTestId('element')
    expect(checkboxElement).toBeInTheDocument();
  });

  it('should render correct checkbox based on checkbox porps', () => {
    const checkboxProps = {
      checkboxSize: CheckboxSize.Small,
      checkboxColor: CheckboxColor.Red
    }
    const wrapper = render(<MyCheckbox data-testid="element"
      {...checkboxProps}
      disabled
      className='checkbox-test'
    />)
    const checkboxElement = wrapper.getByTestId('element')
    expect(checkboxElement).toBeInTheDocument();
    expect(checkboxElement?.tagName).toBe('INPUT')
    expect(checkboxElement).toHaveClass('checkbox', 'checkbox-red', 'checkbox-sm', 'checkbox-test')
  })

})