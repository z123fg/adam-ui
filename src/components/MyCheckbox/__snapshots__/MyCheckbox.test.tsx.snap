// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test MyCheckbox Component should snapshot 1`] = `
<DocumentFragment>
  <input
    class=" checkbox checkbox-sm checkbox-nocolor"
    data-testid="element"
    type="checkbox"
  />
</DocumentFragment>
`;
